### PR TITLE
[ie/orf] Add extractor for sound.orf.at podcasts

### LINKED ISSUE #5265: [orf:radio] add support for sound.orf.at
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

anywhere

### Example URLs

Test link:
https://sound.orf.at/kalender/oe1/20221017/0945

### Provide a description that is worded well enough to be understood

Looks like ORF have a new site, where they bundle all radios.
playlists seem to be stored in m3u files:
https://loopstream01.apa.at/playlist.m3u8?channel=oe1&id=2022-10-17_0820_tl_51_7DaysMon10_1945413.mp3&offset=0&offsetende=2101000&player=sound_v1&referer=sound.orf.at&userid=74f3331c-cb4f-4dc9-af71-27fa95e4aa49

or this one:
https://loopstream01.apa.at/chunklist.m3u8?channel=oe1&id=2022-10-17_0820_tl_51_7DaysMon10_1945413.mp3&offset=0&offsetende=2101000&player=sound_v1&referer=sound.orf.at&userid=74f3331c-cb4f-4dc9-af71-27fa95e4aa49&sessionid=20221017174538314450527239

Links for other domains work just fine https://oe1.orf.at/player/20221017/694819

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
./yt-dlp -v -F "https://sound.orf.at/kalender/oe1/20221017/0945"
[debug] Command-line config: ['-v', '-F', 'https://sound.orf.at/kalender/oe1/20221017/0945']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.10.04 [4e0511f27] (zip)
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python 3.10.7 (CPython 64bit) - macOS-12.6-arm64-arm-64bit 
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: phantomjs 2.1.1
[debug] Optional libraries: sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1739 extractors
[debug] [generic] Extracting URL: https://sound.orf.at/kalender/oe1/20221017/0945
[generic] 0945: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] 0945: Extracting information
[debug] Looking for embeds
ERROR: Unsupported URL: https://sound.orf.at/kalender/oe1/20221017/0945
Traceback (most recent call last):
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1553, in __extract_info
    ie_result = ie.extract(url)
  File "./yt-dlp/yt_dlp/extractor/common.py", line 674, in extract
    ie_result = self._real_extract(url)
  File "./yt-dlp/yt_dlp/extractor/generic.py", line 2755, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://sound.orf.at/kalender/oe1/20221017/0945
```

### PR DESCRIPTION
**Pull request by Esokrates**
### Add orf extractor for podcasts

This adds support for downloading podcasts from `https://sound.orf.at/podcast/sender`.

Partly Fixes #5265

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🎙️🇦🇹➕

Add a new extractor class `ORFPodcastIE` for ORF podcasts in the `orf.py` module and register it in the `_extractors.py` module. This allows yt-dlp to download audio programs from the Austrian public broadcaster.

> _New `ORFPodcastIE`_
> _Extracts audio from ORF_
> _Listen in winter_

### Walkthrough
*  Add support for ORF podcasts with a new extractor class ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R1419), [link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-9434d3584a6d02a95cf46967e794cf7d024be50957816605e711c18bd646234fR337-R372))
   * Define the `ORFPodcastIE` class in `orf.py` that inherits from `InfoExtractor` ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-9434d3584a6d02a95cf46967e794cf7d024be50957816605e711c18bd646234fR337-R372))
   * Implement the `_VALID_URL` pattern to match the ORF podcast URLs ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-9434d3584a6d02a95cf46967e794cf7d024be50957816605e711c18bd646234fR337-R372))
   * Implement the `_TESTS` list to provide an example URL and expected metadata ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-9434d3584a6d02a95cf46967e794cf7d024be50957816605e711c18bd646234fR337-R372))
   * Implement the `_real_extract` method to download the JSON data from the ORF API and extract the audio URL, title, description, duration, and series name ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-9434d3584a6d02a95cf46967e794cf7d024be50957816605e711c18bd646234fR337-R372))
   * Use the `clean_html` helper function to remove HTML tags from the description ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-9434d3584a6d02a95cf46967e794cf7d024be50957816605e711c18bd646234fR337-R372))
   * Register the new extractor class in the `_extractors.py` module ([link](https://github.com/yt-dlp/yt-dlp/pull/8486/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R1419))

</details>

### CODE REVIEW

File: yt_dlp/extractor/orf.py [hunk 1]
```diff
@@ -334,6 +334,42 @@ def _real_extract(self, url):
...
+            f'https://audioapi.orf.at/radiothek/api/2.0/podcast/{station}/{show}/{show_id}', show_id)
+
+        return {
+            'id': show_id,
+            'url': data.get('payload').get('enclosures')[0]['url'],
+            'ext': 'mp3',
+            'title': data.get('payload').get('title'),
+            'description': clean_html(data.get('payload').get('description')),
+            'duration': data.get('payload').get('duration') / 1000,
+            'series': data.get('payload').get('podcast').get('title'),
```
Comment (seproDev) [inline yt_dlp/extractor/orf.py:369]:
```suggestion
            'ext': 'mp3',
            **traverse_obj(data, ('payload', {
                'url': ('enclosures', 0, 'url'),
                'title': 'title',
                'description': ('description', {clean_html}),
                'duration': ('duration', {functools.partial(float_or_none, scale=1000)}),
                'series': ('podcast', 'title'),
            })),
```
Make sure to also add `from ..utils.traversal import traverse_obj`

File: yt_dlp/extractor/orf.py [hunk 2]
```diff
@@ -334,6 +334,42 @@ def _real_extract(self, url):
             self._entries(data, station or station2), show_id, data.get('title'), clean_html(data.get('subtitle')))
 
 
+class ORFPodcastIE(InfoExtractor):
+    IE_NAME = 'orf:podcast'
+
+    _STATION_RE = 'noe|wie|bgl|ooe|stm|ktn|sbg|tir|vbg|oe3|oe1|tv'
+
+    _VALID_URL = f'https?://sound.orf.at/podcast/(?P<station>{_STATION_RE})/(?P<show>[a-zA-Z0-9_-]+)/(?P<id>[a-zA-Z0-9_-]+)'
```
Comment (seproDev) [inline yt_dlp/extractor/orf.py:342]:
```suggestion
    _VALID_URL = rf'https?://sound\.orf\.at/podcast/(?P<station>{_STATION_RE})/(?P<show>[a-zA-Z0-9_-]+)/(?P<id>[a-zA-Z0-9_-]+)'
```

File: yt_dlp/extractor/orf.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/orf.py:369]: also add `'vcodec': 'none'`

File: yt_dlp/extractor/orf.py (hunk 2 above)
Comment (bashonly) [reply yt_dlp/extractor/orf.py:342]:
```suggestion
    _VALID_URL = rf'https?://sound\.orf\.at/podcast/(?P<station>{_STATION_RE})/(?P<show>[\w-]+)/(?P<id>[\w-]+)'
```

File: yt_dlp/extractor/orf.py (hunk 1 above)
Comment (Esokrates) [reply yt_dlp/extractor/orf.py]: Addressed, thanks!

File: yt_dlp/extractor/orf.py (hunk 2 above)
Comment (Esokrates) [reply yt_dlp/extractor/orf.py]: Addressed, thanks!

File: yt_dlp/extractor/orf.py [hunk 3]
```diff
@@ -334,6 +335,45 @@ def _real_extract(self, url):
...
 
 
+class ORFPodcastIE(InfoExtractor):
+    IE_NAME = 'orf:podcast'
+
+    _STATION_RE = 'noe|wie|bgl|ooe|stm|ktn|sbg|tir|vbg|oe3|oe1|fm4|tv'
+
+    _VALID_URL = rf'https?://sound\.orf\.at/podcast/(?P<station>{_STATION_RE})/(?P<show>[\w-]+)/(?P<id>[\w-]+)'
+
+    _TESTS = [{
```
Comment (bashonly) [inline yt_dlp/extractor/orf.py:345]: for the stations, let's use an alphabetized tuple to make it easier to read/maintain
```suggestion
    IE_NAME = 'orf:podcast'
    _STATION_RE = '|'.join(map(re.escape, (
        'bgl', 'fm4', 'ktn', 'noe', 'oe1', 'oe3',
        'ooe', 'sbg', 'stm', 'tir', 'tv', 'vbg', 'wie')))
    _VALID_URL = rf'https?://sound\.orf\.at/podcast/(?P<station>{_STATION_RE})/(?P<show>[\w-]+)/(?P<id>[\w-]+)'
    _TESTS = [{
```

File: yt_dlp/extractor/orf.py [hunk 4]
```diff
@@ -334,6 +335,45 @@ def _real_extract(self, url):
...
+    }]
+
+    def _real_extract(self, url):
+        station, show, show_id = self._match_valid_url(url).group('station', 'show', 'id')
+        data = self._download_json(
+            f'https://audioapi.orf.at/radiothek/api/2.0/podcast/{station}/{show}/{show_id}', show_id)
+
+        return {
+            'id': show_id,
+            'ext': 'mp3',
```
Comment (bashonly) [inline yt_dlp/extractor/orf.py:365]: Are we sure that it is always an mp3?
In the API response JSON there is a mimetype given in the `payload.enclosures.0.type` field

So you could leave `'ext': 'mp3'` here as a default, but then also add

```diff
             **traverse_obj(data, ('payload', {
                 'url': ('enclosures', 0, 'url'),
+                'ext': ('enclosures', 0, 'type', {mimetype2ext}),
```
to extract actual extension when available

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/orf.py (hunk 4 above)
Comment (Esokrates) [reply yt_dlp/extractor/orf.py:365]: It was an mp3 for the various links I tested, but I fully agree :-). Addressed this already.

File: yt_dlp/extractor/orf.py [hunk 5]
```diff
@@ -20,7 +20,9 @@
     unified_strdate,
     unsmuggle_url,
     url_or_none,
+    mimetype2ext,
```
Comment (bashonly) [inline yt_dlp/extractor/orf.py:23]: can you run `isort` on this please?

**Review by bashonly — APPROVED** [review]
LGTM

File: yt_dlp/extractor/orf.py (hunk 5 above)
Comment (Esokrates) [reply yt_dlp/extractor/orf.py]: Done :-)

**Review by seproDev — APPROVED** [review]

**Pull request merged**